src/target/riscv/riscv-013.c

```diff
@@ -2630,15 +2654,58 @@ static int sb_write_address(struct target *target, target_addr_t address,
 static int batch_run(struct target *target, struct riscv_batch *batch)
 {
 	RISCV_INFO(r);
-	const int result = riscv_batch_run(batch, /*resets_delays*/  r->reset_delays_wait >= 0,
+	riscv_batch_add_nop(batch);
+	const int result = riscv_batch_run_from(batch, 0,
+			/*resets_delays*/  r->reset_delays_wait >= 0,
 			r->reset_delays_wait);
-	/* TODO: `finished_scans` should be the number of scans that have
-	 * finished, not the number of scans scheduled. */
+	/* TODO: To use `riscv_batch_finished_scans()` here, it is needed for
+	 * all scans to not discard input, meaning
+	 * "riscv_batch_add_dm_write(..., false)" should not be used. */
 	const size_t finished_scans = batch->used_scans;
 	decrement_reset_delays_counter(target, finished_scans);
 	return result;
 }
 
+/* It is expected that during creation of the batch
+ * "riscv_batch_add_dm_write(..., false)" was not used. */
+static int batch_run_timeout(struct target *target, struct riscv_batch *batch)
+{
+	RISCV013_INFO(info);
+	size_t finished_scans = 0;
+	const time_t start = time(NULL);
+	riscv_batch_add_nop(batch);
+	int result = ERROR_FAIL;
+	do {
+		RISCV_INFO(r);
+		result = riscv_batch_run_from(batch, finished_scans,
+				/*resets_delays*/  r->reset_delays_wait >= 0,
+				r->reset_delays_wait);
+		const size_t new_finished_scans = riscv_batch_finished_scans(batch);
+		assert(new_finished_scans >= finished_scans);
+		decrement_reset_delays_counter(target, new_finished_scans - finished_scans);
+		finished_scans = new_finished_scans;
+		if (result != ERROR_OK)
+			return result;
+		if (!riscv_batch_dmi_busy_encountered(batch)) {
+			assert(finished_scans == batch->used_scans);
+			return ERROR_OK;
+		}
+		increase_dmi_busy_delay(target);
+		riscv_batch_update_idle(batch, info->dmi_busy_delay);
+	} while (time(NULL) - start < riscv_command_timeout_sec);
+	assert(result == ERROR_OK);
+	assert(riscv_batch_dmi_busy_encountered(batch));
+	/* Reset dmi_busy_delay, so the value doesn't get too big. */
+	LOG_TARGET_DEBUG(target, "dmi_busy_delay is reset.");
+	info->dmi_busy_delay = 0;
+	/* TODO: Introduce a separate timeout for this. */
+	LOG_TARGET_ERROR(target, "DMI operation didn't complete in %d seconds. "
+			"The target is either really slow or broken. You could increase "
+			"the timeout with riscv set_command_timeout_sec.",
+			riscv_command_timeout_sec);
+	return ERROR_TIMEOUT_REACHED;
+}
```

Comment:
This part of the code is written in a very dense style. Few blank lines will improve readability:

```suggestion
	} while (time(NULL) - start < riscv_command_timeout_sec);
	
	assert(result == ERROR_OK);
	assert(riscv_batch_dmi_busy_encountered(batch));
	
	/* Reset dmi_busy_delay, so the value doesn't get too big. */
	LOG_TARGET_DEBUG(target, "dmi_busy_delay is reset.");
	info->dmi_busy_delay = 0;

	/* TODO: Introduce a separate timeout for this. */
	LOG_TARGET_ERROR(target, "DMI operation didn't complete in %d seconds. "
			"The target is either really slow or broken. You could increase "
			"the timeout with riscv set_command_timeout_sec.",
			riscv_command_timeout_sec);
	return ERROR_TIMEOUT_REACHED;
}
```

Reply:
Addressed.